DAOS-6540 dfuse: Improve the dfuse multi-user documentation. (#11511)
daos-stack/daos: docs/user/multi-user-dfuse.md

Namespace link is created after the container itself is created.

## Controlling data sharing and access.

DFuse itself does not handle permissions checks regardless of options but rather provides file and
directory ownership and permissions bits information to the kernel which then checks if user is
entitled to perform the requested operation.  As such once a multi-user dfuse instance is running it
will serve to the kernel the contents of any containers that the dfuse instance itself can access,
and the kernel will then use standard Unix file/directory permissions to decide if operations are
permitted.  It is therefore possible for users be able to access POSIX data across containers
seamlessly where permissions is granted only at the POSIX file/directory layer.

Consider the following example: dfuse is configured to run in the recommended manner, with two users
Anthony and Berlinda who use it to share data, yet Berlinda does not have ACL permissions to read
Anthony's data - only POSIX permissions.

Example:

### Setup user to serve dfuse and create containers, mount points etc.
```bash

> $ sudo -u dserve dmg --insecure pool create root_pool --size 1g

Reviewer: no --insecure in user-facing docs please...

Author: I thought I'd removed these already.  Gone now.

$ sudo -u dserve daos cont create --type POSIX root_pool root_container

> $ sudo mkdir /crate

Reviewer: rename /crate to /daos ?

Author: We've previously had problems with users getting confused between and server mounts/fileystems etc and the dfuse one which if why I'm shying away from calling a dfuse mount "/daos" but I can change it if you think it's better.

$ sudo chown dserve.dserve /crate
```

### Run dfuse, this should be done via systemd to be automatically mounted at boot time.
```bash
$ sudo -u dserve dfuse --multi-user /crate root_pool root_container
```

### Create a directory for anthony to own, and create a pool for him.
```bash
$ sudo mkdir -m 0700 /crate/anthony
$ sudo chown anthony.anthony /crate/anthony
$ sudo dmg pool create -u anyhony -g anthony anthony_pool --size 1g

> $ sudo -u dserve dmg --insecure pool update-acl anthony_pool -e "A::dserve@:r"

Reviewer: no --insecure in user-facing docs please...

> $ sudo -u dserve dmg --insecure pool update-acl anthony_pool -e "A::dserve@:r"

Reviewer: why "-u dserver" here? dmg must be run as a user who has dmg permissions, not as the dserve user (which i would strongly discourage to have dmg permissions)...

Author: Good catch, presumably this should be root as they created the pool.

> $ sudo -u anthony daos cont create --path /crate/anthony/my-data anthony_pool --type POSIX

Reviewer: don't we need a cont_label ?

Author: No.

$ sudo -u anthony chmod 755 /crate/anthony

> $ sudo -u anthony sh -c "echo hello-world > /crate/anthony/my-data/new-file"

Author: well, the permissions depend on the umask - and that's down to the user to set.

```

### Now read the file.
```bash

> $ sudo -u berlinda cat /crate/anthony/my-data new-file

Reviewer: blank before new-file should be "/" ?

Author: Yes I think so but I haven't updated this in the latest push, I'll go through the workflow again on my system today and verify this.

sudo -u dserve dfuse --multi-user --foreground /crate root_pool root_container

### Create a directory for anthony to own, and create a pool for him.
sudo mkdir -m 0700 /crate/anthony
sudo chown anthony.anthony /crate/anthony
sudo dmg --insecure pool create -u anyhony -g anthony anthony_pool --size 1g
sudo -u dserve dmg --insecure pool update-acl anthony_pool -e "A::dserve@:r"

sudo -u anthony daos cont create --path /crate/anthony/my-data anthony_pool --type POSIX
sudo -u anthony chmod 755 /crate/anthony
sudo -u anthony sh -c "echo hello-world > /crate/anthony/my-data/new-file"

### Now read the file.
sudo -u berlinda cat /crate/anthony/my-data new-file

## Interception library

Use of the interception library with multi-user dfuse is supported with no change in configuration,
however in the case where dfuse and the interception library are being run as different users
then the DAOS pool/container handles will not be shared across process boundaries.  In this case
the interception library will itself call pool connect and container open leading to greater
overhead in opening files.  This change should be seamless to the user.

## plans?

> Kris has the ACL PR which can be used to enforce 

Reviewer 2: (style)  trailing whitespace

> # Multi user dfuse.

Reviewer 3: Do we want a `-` in the docs or no?
```suggestion
# Multi-user dfuse.
```

## Overview

> Multi user dfuse is a operating mode of dfuse where it is able to provide filesystem access to
> multiple users/groups on a node through the same dfuse daemon.  This can be used persistently

Reviewer 3: ```suggestion
Multi user dfuse is an operating mode of dfuse where it is able to provide filesystem access to
multiple users/groups on a node through the same dfuse daemon.  This can be used persistently
```

on a node to provide data sharing and scratch style access or on-demand by specific users to allow
others to access their data.

### Setup

To enable multi user dfuse libfuse needs to be reconfigured as root to allow other users to access
the mount point.  This is a per-node setting to a configuration file in /etc and will apply for all

> users on a node.  Only root can enable this, but once set it applies to all users.

Reviewer 3: Should we go ahead and specify what the config setting is?

Author: Yes, this should probably be in the admin guide as well, I'll update.  There are security implications of this so we do not want to do it via rpm scripts for example and it will require a manual admin step.

Author: Admin guide is updated as part of this PR.

## Recommended use-case

Persistent multi user dfuse instances make sense on login nodes to allow permanent access to DAOS

> for multiple users without the overhead of spinning up per-user instances on login, it allows

Reviewer 3: ```suggestion
for multiple users without the overhead of spinning up per-user instances on login. It allows
```

better use of kernel resources, is easier to administer and more flexibile than per-user dfuse
processes.

To set this up it is assumed that the system admins will create a dedicated unix account for this
and configure systemd on login/compile nodes to start dfuse as part of the boot sequence for the
nodes.  This dedicated user should have a pool and container that are owned by them to
serve as the "root" of the dfuse mount, and systemd should be configured to provide these details
to dfuse at startup.  It is anticipated that this root container will serve only to host links to
other containers owned by different users and will not itself contain significant data.

## Unified Namespace.

### Requirement for setting of ACLs.

> To make use of Unified Namespace across users appropriate ACLs need to be set on both DAOS pools

Reviewer 3: "Unified Namespace across users"?

> To make use of Unified Namespace across users appropriate ACLs need to be set on both DAOS pools

Author: This should be already covered in https://github.com/daos-stack/daos/blob/master/docs/user/filesystem.md#links-into-other-containers although I'm surprised it doesn't specifically say "Unified Namespace" there.

and containers.  DFuse is like any other daos client in this regard and when running multi-user
dfuse then it is likely that the user dfuse is running as does not own all the pools and containers
it will be serving.

ACLs required for pools is the 'r' permission on any pool for the user running
dfuse.  This administrator may choose to set this when creating pools.

It is possible for a user to run multi-user dfuse and to do so purely to expose data in their own

> containers to other users, in this case the user running dfuse is the same as the user owning the

Reviewer 3: ```suggestion
containers to other users. In this case the user running dfuse is the same as the user owning the
```

containers so no modification of ACLs is required.

### Automatic setting of ACLs.

For containers the 'r', 't' and 'w' permissions are required on the container for the user running

> dfuse.  When containers are created though the `daos` command using a path which resides in
> a multi-user dfuse mount then these ACLs are set automaticlaly by the `daos` command and this
> is reported to the user.  (TODO: Check this is reported).

Reviewer 3: 1. Should we address this TODO?
2. Recommend not repeating
3. spelling
```suggestion
dfuse.  When containers are created though the `daos` command using a path which resides in
a multi-user dfuse mount then these ACLs are set automatically and this
is reported to the user.  (TODO: Check this is reported).
```

> If a container is created using a path not backed by multi-user dfure, for example on a per-user
> dfuse instance or where a `path` option is not provided at all then in order for
> anyone to access the container then the user will need to add this permission themselves.  This
> includes the user that owns the pool, without the dfuse users being able to access the container
> it cannot server the contents, even to the container owner.  This applies equally well to containers
> where the Unified Namespace link is created after the container itself is created.

Reviewer 3: ```suggestion
If a container is created using a path not backed by multi-user dfuse - for example, on a per-user
dfuse instance or where a `path` option is not provided at all - then in order for
anyone to access the container the user will need to add this permission themselves.  This
includes the user that owns the pool. Without the dfuse users being able to access the container
it cannot serve the contents, even to the container owner.  This applies equally well to containers
where the Unified Namespace link is created after the container itself is created.
```

> TODO: Check this, create a container single-user, access it via multi-user and check for errors,

Reviewer 3: TODO?

> can 'daos cont set-acl' work on a dfuse-path in the multi-user setup?

Author: Good point, it probably doesn't.  @daltonbohning can you add this to the testing?  Probably worth testing it for both single-user and multi-user separately.

> can 'daos cont set-acl' work on a dfuse-path in the multi-user setup?

Reviewer 3: @ashleypittman Yes, but please create a ticket stating what needs to be tested. I'm not 100% sure what we need to test here

## Controlling data sharing and access.

> DFuse itself does not handle permissions checks regardless of options but rather provides file
> and directory ownership and permissions bits information to the kernel which then checks if user
> is entitedled to perform the requested operation.  As such once a multi-user dfuse instance is
> running then it will serve to the kernel the contents of any containers that the dfuse instance
> itself can access, and the kernel will then use standard Unix file/directory permissions to decide
> if operations are permitted.  It is therefore possible for users be able to access POSIX data across
> containers seamlessly where permissions is granted only at the POSIX file/directory layer.

Reviewer 3: ```suggestion
DFuse itself does not handle permissions checks regardless of options but rather provides file
and directory ownership and permissions bits information to the kernel which then checks if user
is entitled to perform the requested operation.  As such once a multi-user dfuse instance is
running it will serve to the kernel the contents of any containers that the dfuse instance
itself can access, and the kernel will then use standard Unix file/directory permissions to decide
if operations are permitted.  It is therefore possible for users be able to access POSIX data across
containers seamlessly where permissions is granted only at the POSIX file/directory layer.
```

> Consider the following example, in this case dfuse is configured and run in the recommended manner,
> with two users Anthony and Berlinda who use it to share data, yet Berlinda does not have ACL
> permissions to read Anthonys data, only POSIX permissions.

Reviewer 3: ```suggestion
Consider the following example: dfuse is configured to run in the recommended manner,
with two users Anthony and Berlinda who use it to share data, yet Berlinda does not have ACL
permissions to read Anthony's data - only POSIX permissions.
```

Example:

### Setup user to serve dfuse and create containers, mount points etc.
sudo -u dserve dmg --insecure pool create root_pool --size 1g
sudo -u dserve daos cont create --type POSIX root_pool root_container
sudo mkdir /crate

> sudo chown dserve.dserve /crate

Reviewer 3: Is code format supported here? I.e. wrapping these commands with three back ticks: ```

### Run dfuse, this should be done via systemd to be automatically mounted at boot time.

> sudo -u dserve dfuse --multi-user --foreground /crate root_pool root_container

Reviewer 4: why foreground?

Author: Removed.  I was working though these steps in the terminal to ensure they were correct and this and the --insecure were carried over from that.

### Create a directory for anthony to own, and create a pool for him.
sudo mkdir -m 0700 /crate/anthony
sudo chown anthony.anthony /crate/anthony

> sudo dmg --insecure pool create -u anyhony -g anthony anthony_pool --size 1g

Reviewer 4: please remove insecure mode. it should not be used in production, and it might confuse readers that this is required in this case.

> sudo mkdir -m 0700 /crate/anthony
> sudo chown anthony.anthony /crate/anthony
> sudo dmg --insecure pool create -u anyhony -g anthony anthony_pool --size 1g
> sudo -u dserve dmg --insecure pool update-acl anthony_pool -e "A::dserve@:r"

Reviewer 3: ```suggestion
sudo mkdir -m 0700 /crate/anthony
sudo chown anthony.anthony /crate/anthony
sudo dmg --insecure pool create -u anthony-g anthony anthony_pool --size 1g
sudo -u dserve dmg --insecure pool update-acl anthony_pool -e "A::dserve@:r"
```

Author: The only change I see here is removing a space before the -g which is going to break this isn't it?

Reviewer 3: Oops. Should be
`anyhony -g` ->
`anthony -g`

sudo -u anthony daos cont create --path /crate/anthony/my-data anthony_pool --type POSIX
sudo -u anthony chmod 755 /crate/anthony
sudo -u anthony sh -c "echo hello-world > /crate/anthony/my-data/new-file"

### Now read the file.

> sudo -u berlinda cat /crate/anthony/my-data new-file

Reviewer 4: could use some refinement on the example by adding some comments on some commands.
had to read a few times to grok

> sudo -u berlinda cat /crate/anthony/my-data new-file

Reviewer 4: should mention that while Berlinda can read the file, she has no access to the container itself outside of this dfuse mount. meaning cannot use  DFS API for example to read that file.

## Interception library

Use of the interception library with multi-user dfuse is supported with no change in configuration,
however in the case where dfuse and the interception library are being run as different users
then the DAOS pool/container handles will not be shared across process boundaries.  In this case
the interception library will itself call pool connect and container open leading to greater
overhead in opening files.  This change should be seamless to the user.

## plans?

Kris has the ACL PR which can be used to enforce 

> https://github.com/daos-stack/daos/pull/11359

Reviewer 3: Do we want to keep this in the docs?

Author: No, note-to-self.

## Overview

Multi user dfuse is an operating mode of dfuse where it is able to provide filesystem access to
multiple users/groups on a node through the same dfuse daemon.  This can be used persistently on a

> node to provide data sharing and scratch style access or on-demand by specific users to allow others

Reviewer 3: ```suggestion
node to provide data sharing and scratch-style access or on-demand by specific users to allow others
```

## Interception library

Use of the interception library with multi-user dfuse is supported with no change in configuration,

> however in the case where dfuse and the interception library are being run as different users then

Reviewer 3: ```
. However,
```
or
```
; however,
```
